backend/corpora/dataset_processing/upload.py

```diff
@@ -0,0 +1,139 @@
+        manager.session.commit()
+
+
+def updater(upload_uuid: str, tracker: ProgressTracker, frequency: float = 3):
+    """
+    Update the progress of an upload to the database using the tracker.
+
+    :param upload_uuid: The uuid of the upload_progress row.
+    :param tracker: Tracks information about the progress of the upload.
+    :param frequency: The frequency in which the database is updated
+    :return:
+    """
+
+    def _update():
+        progress = tracker.progress()
+        if progress > 1:
+            tracker.stop_uploader.set()
+            processing_status = {
+                DbDatasetProcessingStatus.upload_progress: progress,
+                DbDatasetProcessingStatus.upload_message: "The file size, does not match the size of the upload.",
+                DbDatasetProcessingStatus.upload_status: UploadStatus.FAILED,
+            }
+        elif progress == 1 and tracker.stop_updater.isSet():
+            processing_status = {
+                DbDatasetProcessingStatus.upload_progress: progress,
+                DbDatasetProcessingStatus.upload_status: UploadStatus.UPLOADED,
+            }
+        else:
+            processing_status = {DbDatasetProcessingStatus.upload_progress: progress}
+        processing_status_updater(upload_uuid, processing_status)
+
+    try:
+        while not tracker.stop_updater.wait(frequency):
+            _update()
+        _update()  # Make sure the progress is update once the upload is complete
+    finally:
+        tracker.stop_uploader.set()
+
+
+def upload(dataset_uuid: str, url: str, local_path: str, file_size: int, chunk_size: int = 10 * MB, update_frequency=3):
+    """
+    Upload a file from a url and update the processing_status upload fields in the database
+
+    :param dataset_uuid: The uuid of the dataset the upload will be associated with.
+    :param url: The URL of the file to be uploaded.
+    :param local_path: The local name of the file be uploaded
+    :param file_size: The size of the file in bytes.
+    :param chunk_size: The size of downloaded data to copy to memory before saving to disk.
+    """
+    with db_session_manager() as mananger:
+        processing_status = Dataset.get(dataset_uuid).processing_status
+        processing_status.upload_status = UploadStatus.UPLOADING
+        processing_status.upload_progress = 0
+        mananger.commit()
+        upload_uuid = processing_status.id
+    progress_tracker = ProgressTracker(file_size)
+    progress_thread = threading.Thread(
+        target=updater,
+        kwargs=dict(upload_uuid=upload_uuid, tracker=progress_tracker, frequency=update_frequency),
+    )
+    progress_thread.start()
+    upload_thread = threading.Thread(
+        target=uploader, kwargs=dict(url=url, local_path=local_path, tracker=progress_tracker, chunk_size=chunk_size)
+    )
+    upload_thread.start()
+    upload_thread.join()
```

Comment:
can you add a comment about why join is used here?

Reply:
added